_posts/2017-10-04-scalamacros.md

```diff
@@ -0,0 +1,299 @@
+his collaborators to establish a non-experimental and portable macro system for
+Scala.
+With Eugene's blessing, I will be taking the lead on behalf of the Scala Center
+to develop this project in close collaboration with the Scala community, the
+Dotty team at EPFL, Scala compiler team at Lightbend and the IntelliJ Scala
+Plugin team at Jetbrains.
+This initiative follows [SCP-014], a proposal that was approved two weeks ago
+with an overwhelming majority of the Scala Center Advisory Board.
+
+## Brief history
+Before diving into the roadmap of this new exciting development, I'd like to
+briefly summarize and recognize the efforts that have been made so far towards
+establishing a standard, non-experimental macro system for Scala.
+I'll try to keep it short here, for a more comprehensive coverage please refer
+to the list of [papers written by Eugene Burmako][papers], the founder of Scala
+Macros, in particular his PhD thesis [Unification of Compile-Time and Runtime
+Metaprogramming in Scala][thesis].
+
+If you want the TL;DR, see [next steps](#next-steps).
+
+### v1: scala.reflect
+
+Scala.reflect-based macros are an integral part of the Scala 2.x ecosystem.
+Well-known libraries like ScalaTest, Sbt, Spark, Circe, Slick, Shapeless,
+Spire and others use scala.reflect macros to achieve previously unreachable
+standards of expressiveness, type safety and performance.
+
+Unfortunately, scala.reflect-based macros have also gained notoriety as an
+arcane and brittle technology.
+The most common criticisms of scala.reflect is its sub-par tooling support
+and non-portable metaprogramming API based on Scala 2.x compiler internals.
+Even five years after their introduction, scala.reflect macros still can't
+expand in IntelliJ, leading to proliferation of spurious red squiggles -
+sometimes in pretty simple code.
+
+Quoting Eugene Burmako, the author of Scala macros, in [SIP-16] on
+self-cleaning macros
+
+> While trying to fix these problems via evolutionary changes to the current
+> macro system, we have realized that most of them are caused by the decision
+> to use scala.reflect as the underlying metaprogramming API. Modelled after
+> compiler internals, scala.reflect inherits many of its peculiar design
+> choices. Extensive use of desugarings and existence of multiple independent
+> program representations may have worked well for compiler development, but
+> they turned out to be inadequate for a public API.
+
+As a result of these known limitations, the language committee has decided to
+retire the scala.reflect-based macro system.
+Another justification for retiring the scala.reflect-based macro system was
+that a new macro system based on Scalameta was "just around the corner".
+
+### v2: scala.meta
+
+The [Scalameta] project was founded to become a better macro system for Scala,
+with the vision to replace scala.reflect as the de-facto metaprogramming
+toolkit for Scala.
+
+With Scalameta, we managed to support macro annotations in
```

Comment:
It's probably also a good idea to note that annotation macros aren't supported in Scala 2.x without the paradise plugin (not everyone will necessarily click on the link). In reality, I'd argue that annotation macros really actually aren't supported in Scala 2.x because you can't use them with the production compiler alone...

Reply:
Good point, updated.